internal/crypto/cipher.go

```diff
@@ -0,0 +1,208 @@
+package crypto
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"fmt"
+	"hash"
+
+	"github.com/optable/match/internal/util"
+	"github.com/zeebo/blake3"
+)
+
+/*
+Various cipher suite implementation in golang
+*/
+
+const (
+	GCM = iota
+	XORBlake3
+
+	nonceSize = 12 //aesgcm NonceSize
+)
+
+// PseudorandomCode is implemented as follows:
+// C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
+// secretKey is a 16 byte slice for AES-128
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCode(aesBlock cipher.Block, src []byte) (dst []byte) {
+	dst = make([]byte, aes.BlockSize*4)
+	// effectively pad src
+	copy(dst, src)
+
+	// encrypt
+	aesBlock.Encrypt(dst[:aes.BlockSize], dst[:aes.BlockSize])
+	if len(src) <= aes.BlockSize {
```

Comment:
I am confused by this function. Do we early exit for optimization or because we must not encrypt padding blocks? What's confusing is that we still encrypt padding bytes inside a block.

I'm under the impression that we never pass something that's smaller than a block anyway so I would just encrypt everything if it has no consequence. I don't think we're really improving things by doing these branches if it's rare that we get less than 4 blocks in src.

Reply:
`aesBlock.Encrypt` operates on a single block (16 bytes). There is not point to encrypting padding blocks but we can't avoid encrypting padding within a block. We exit early for optimization as each call to this is shown to be costly when profiling. I agree however, that the branches only add overhead when our src is already 64 bytes.

Reply:
Have you confirmed that doing early exits was an improvement?

Reply:
@xcapaldi : I have run a quick benchmark myself. Turns out that it's very similar for the 64 bytes case to run with or without branches, as expected. It's about 25% more expensive to encrypt all of the blocks. I think some of it is explained by the virtual calls through the cipher.block interface which is not optimized by the compiler. The AESENC instructions are never inlined in the PseudoRandomCode. I am not able to demonstrate it though and I don't think we'll be able to make this much better for now.